### ISSUE TITLE
Allow CSS-like syntax

### ISSUE THREAD

**Issue by denkwai** [curvedmark/roole#36]
It would be really nice if I could use braces to define styles. Many people would prefer this code style

```
$width = 400px
.block {
    width: $width
    height: 200px

    .inner {
        $width = 200px
        width: $width
    }
}
```

instead of this one

```
$width = 400px
.block
    width: $width
    height: 200px

    .inner
        $width = 200px
        width: $width
```

I like the Stylus way of solving this problem ­- it allows both indent-based and braces-based syntax. Why not implement it here?

**Comment by curvedmark** [comment]
I know this one will come.

To be honest, when I first started to design Roole, it's a superset of CSS, but later I gave it up.

Ask yourself this: why do I prefer the CSS style? I can think of three possible reasons:
1. So adapting old CSS code would be easy. I just have to copy & paste old code, add a few variables if possible and I'm done.
2. I have more freedom on how to format the code. I can use whichever style that is [introduced here](http://css-tricks.com/different-ways-to-format-css/)
3. I'm more familiar with CSS, so if this language use the CSS syntax, I can learn it more easily.

These are my arguments against them:
1. It's the job of a CSS to Roole conversion tool. You should first convert your CSS code to Roole using an automatic tool, and then you can copy & paste again.
2. Notice the acceptable coding styles are actually limited. Roole does its best to allow all kinds of coding styles. You can put all selectors in a single line, you can put all properties in a single line, you can put multiple values for a single property in multiple lines (in the future). So you won't lose the freedom.
3. No matter which type of syntax the language chooses, you always have to learn the new language constructs that are not in the CSS (variables, `@if`, `@extend`, etc), which define most part of your learning curve. I can hardly  believe by replacing braces with indentation someone would have difficulties in understanding the code. Just look at the two snippets your provided, are they really that different? I personally find the latter to be slightly cleaner and thus slightly easier to read and understand.

I also ask myself: why do I prefer the indentation style? And the answers are:
1. So I can get rid of code like this:
   
   ```
           }
       }
   }
   ```
2. I already formated my code nicely, and the language is able to understand my code structure by looking at the formation. Why would I want to add noises like braces to the code? If it's not necessary, it shouldn't be there.

I don't want to allow two syntaxes simultaneously because that would complicate the parser so much. If I really have to do that, I would take the Sass way: splitting the language into two separate syntaxes.

But I'm open to any suggestions, arguments, etc, that you might have.

**Comment by denkwai** [comment]
Actually, Stylus is limited in braces-based syntax: in fact, it copies the intend syntax (if you don't make intends inside the braces, parser fails), but allows to use braces in two ways:

```
.class {
    <some style>
}
```

and 

```
.class
{
    <some style>
}
```

I don't think this way of parsing will complicate parser much. Please, correct me if I'm wrong.

**Comment by curvedmark** [comment]
So the syntax brings noises from CSS but not the freedom that should come with it?

**Comment by denkwai** [comment]
The syntax brings a visual effect, that makes code reading and writing more comfortable in some cases.
I really like indent-based markup, but many people are familiar with braces-based code partition because they have been practicing programming on C-based syntax languages, not Python or Ruby.
If you don't want to implement this syntax here, I think you should create a fork of this project, as you proposed. That would be a great improvement for the users of your software.

**Comment by curvedmark** [comment]
Closing for now until more people think the CSS syntax is superior.

Please don't simply :+1: here, provide a concrete reason as to why you like the CSS syntax more.

**Comment by jamiebuilds** [comment]
As far as I've seen, more people prefer following the SCSS syntax over the SASS syntax (I've almost never seen SASS syntax out in the wild, although SCSS is everywhere). Regardless of reason, I'd hope that you'd target developer happiness.

**Here are my personal reasons for using SCSS over SASS:**
1. I work with a lot of engineers from different backgrounds (mostly backend engineers), and we follow a strict style guide, and I've made that style guide based largely on CSS so that when they need to look something up it isn't foreign to them. Not everyone is familiar with whitespace significant languages, so it leads to confusion. It also presents a harder learning curve if they've only ever worked in CSS.
2. Trying to follow normal CSS as much as possible helps me write DRYer smaller code. I see people going wild with preprocessor selectors that bulk up code and lead to confusion.
3. In the upcoming CSS Level 4 Spec, they are adding several new selectors like [the Subject! selector](http://dev.w3.org/csswg/selectors4/#subject), [reference /for/ combinators](http://dev.w3.org/csswg/selectors4/#idref-combinators), [:matches(selector)](http://dev.w3.org/csswg/selectors4/#matches), etc. CSS is picking up a lot of new complex selectors. Seeing all of this discussed will lead to confusion of how and if it can be done using an entirely different syntax.
4. It's difficult to enforce best practices when people are seeing code formatted in two entirely different ways, I end up fixing a lot of peoples code.
5. Call me old fashioned but I'm comfortable with the CSS syntax, although I do want some help writing it, I do not want to abandon it completely.

I know they can seem like petty reasons when being viewed from a front end perspective, but we have to work with the people around us and I don't want to put up any walls for them that I'm going to constantly help them with any more than I have to. I'd rather discuss the individual properties and how they work rather than waste my time on syntax just to save a few hundred characters of writing. I like what the SCSS language has done, and at this point I plan on continuing using it when working on a team. 

I personally wish that Roole would just ignore the brackets when writing selectors.. I feel like that wouldn't be to hard to implement, and there wouldn't be a need for different file extensions like SASS/SCSS.

**Comment by curvedmark** [comment]
After thinking about it again, supporting two syntax simultaneously might not complicate the parser that much, I will give it a try when I got the time.

**Comment by jamiebuilds** [comment]
:smile_cat: :+1:

**Comment by curvedmark** [comment]
Landed in f4161f1a3d8b58d14147c50c696fe9bdf0448935

Use the syntax suggested by @inkwai which is how Stylus works.
